plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -645,47 +717,79 @@ private static ReaderPageSource createParquetPageSource(
             TupleDomain<ColumnDescriptor> parquetTupleDomain = getParquetTupleDomain(descriptorsByPath, effectivePredicate);
             Predicate parquetPredicate = buildPredicate(requestedSchema, parquetTupleDomain, descriptorsByPath, UTC);
 
+            long nextStart = 0;
+            ImmutableList.Builder<Long> blockStarts = ImmutableList.builder();
             List<BlockMetaData> blocks = new ArrayList<>();
             for (BlockMetaData block : parquetMetadata.getBlocks()) {
                 long firstDataPage = block.getColumns().get(0).getFirstDataPageOffset();
                 if (start <= firstDataPage && firstDataPage < start + length &&
                         predicateMatches(parquetPredicate, block, dataSource, descriptorsByPath, parquetTupleDomain)) {
                     blocks.add(block);
+                    blockStarts.add(nextStart);
                 }
+                nextStart += block.getRowCount();
             }
 
             MessageColumnIO messageColumnIO = getColumnIO(fileSchema, requestedSchema);
             ParquetReader parquetReader = new ParquetReader(
                     Optional.ofNullable(fileMetaData.getCreatedBy()),
                     messageColumnIO,
                     blocks,
-                    Optional.empty(),
+                    blockStarts.build(),
                     dataSource,
                     UTC,
                     memoryContext,
                     options);
 
+            ImmutableList.Builder<ConstantPopulatingPageSource.ColumnType> columns = ImmutableList.builder();
+            int parquetSourceChannel = 0;
+
             ImmutableList.Builder<Type> trinoTypes = ImmutableList.builder();
             ImmutableList.Builder<Optional<Field>> internalFields = ImmutableList.builder();
+            ImmutableList.Builder<Boolean> rowIndexChannels = ImmutableList.builder();
             for (int columnIndex = 0; columnIndex < readColumns.size(); columnIndex++) {
                 IcebergColumnHandle column = readColumns.get(columnIndex);
-                org.apache.parquet.schema.Type parquetField = parquetFields.get(columnIndex);
-
-                Type trinoType = column.getBaseType();
-
-                trinoTypes.add(trinoType);
-
-                if (parquetField == null) {
+                if (column.isIcebergIsDeletedMetadataColumn()) {
```

Comment:
This made me think whether we should instead have the method name simply as `isDeletedMetadataColumn()` since the column handle is already an `IcebergColumnHandle`.